[export] don't use unbacked_renamings in export

pianpwk opened the pull request:
Plan: avoid the use of unbacked renamings, and introduce a pass run in `_produce_aten_artifact` that recomputes unbacked bindings. Decided to do this because in we don't serialize unbacked renamings (or any ShapeEnv state), so this used to compose poorly with de/serialization. This hopefully establishes the invariant that the unbacked binding keys are always in sync with the example values (i.e. same indices, and removed if the symbol is replaced / specialized).

For de/serialization, we don't stored unbacked bindings, and just rerun the pass.

Involved a refactor of compute_unbacked_bindings.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

avikchaudhuri reviewed (commented):
As I look at this, requesting a couple more tests meanwhile:
1. (public) Can you add this test: https://github.com/pytorch/pytorch/pull/146115/files#diff-4a060a24e1f81389eab7390d434dddf919af50aa8fda6cbd81e182a53bd9328e

2. (internal) Check that these tests continue to pass (were broken by my previous attempt): https://www.internalfb.com/tasks/?t=214558149

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -3854,6 +3858,22 @@ def foo_unbacked_fake_impl(a, b):
+        # collect bound symbols
+        bound = set()
+        for node in ep.graph.nodes:
+            bound.update(node.meta.get("unbacked_bindings", {}))
```
this is supposed to add the symbol names, right?

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -3854,6 +3858,22 @@ def foo_unbacked_fake_impl(a, b):
+            bound.update(node.meta.get("unbacked_bindings", {}))
+
+        # check ShapeEnv counters compared to binding indices
+        shape_env = detect_fake_mode(
```
I think there's another util to directly detect fake mode from a graph module / ep, which is also used elsewhere in our code. Maybe switch to that?

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -3854,6 +3858,22 @@ def foo_unbacked_fake_impl(a, b):
+            [node.meta.get("val") for node in ep.graph.nodes]
+        ).shape_env
+        next_index = next(shape_env.unbacked_symint_counter)
+        for symbol in bound:
```
Curious to see what happens when we have a `torch._check` or some kind of assert that makes the unbacked symbol go to a constant or backed symint. The test I suggested should have some of that...

avikchaudhuri replied:
Oh maybe the next test `test_replaced_unbacked_bindings` is for that?

avikchaudhuri commented on test/export/test_serialize.py:
```diff
@@ -786,6 +786,49 @@ def forward(self, a, b, c):
+            torch.tensor(4),
+            torch.tensor(5),
+        )
+        ep = torch.export.export(M(), inps, strict=True).run_decompositions()
```
`strict=True` is intentional? Can we remove it?

pianpwk replied:
Yeah, the renaming happens in strict with the AOTAutograd run

avikchaudhuri reviewed (commented):
Still looking...

avikchaudhuri commented on torch/export/_trace.py:
```diff
@@ -462,6 +490,7 @@ def _produce_aten_artifact(
     # Overwrite output specs afterwards.
     flat_fake_args = pytree.tree_leaves((fake_args, fake_kwargs))
     gm, graph_signature = apply_runtime_assertion_pass(gm, graph_signature)
+    _replace_unbacked_bindings(gm)
```
add comment on what we're doing here

avikchaudhuri reviewed (approved):
looks great, address comments before landing

pianpwk commented on test/export/test_serialize.py:
```diff
@@ -786,6 +786,49 @@ def forward(self, a, b, c):
+        bound = set()
+        for old_node, new_node in zip(ep.graph.nodes, loaded_ep.graph.nodes):
+            self.assertTrue(
+                not (
```
**facepalm**

pianpwk commented on torch/_export/serde/serialize.py:
```diff
@@ -1879,6 +1878,18 @@ def deserialize_graph(self, serialized_graph: Graph) -> torch.fx.Graph:
+                (val := node.meta.get("val")) is not None
+                and (
+                    unbacked_bindings := symbolic_shapes._free_unbacked_symbols_with_path(
+                        val, (), shape_env=self.shape_env, pending=self.unbacked_symbols, simplify=True
```
If deserialization is correct, then yeah we should be able to confirm that. the `pending/self.unbacked_symbols` set is built up from parsing the sympy expressions found in the node example values. But it's built up comprehensively first, then consumed gradually across each node.

pianpwk commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -940,6 +940,162 @@ def get(self, o: int) -> int:
         return o // self.divisor
 
 
+def _free_unbacked_symbols_with_path(
```
there were some minor changes, but tries to restore behavior for compile. gonna comment

pianpwk commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -940,6 +940,162 @@ def get(self, o: int) -> int:
+        unbacked = lhs if lhs in pending else rhs
+        divisor: Union[int, SymInt] = (
+            int(coeff)
+            if shape_env and isinstance(coeff, sympy.Integer)
```
added these `if shape_env` checks for export case

pianpwk commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -940,6 +940,162 @@ def get(self, o: int) -> int:
+            torch.sparse_csc,
+            torch.sparse_bsr,
+            torch.sparse_bsc,
+        ]:
```
added this check to not call stride() on such tensors

pianpwk commented on torch/_export/serde/serialize.py:
```diff
@@ -2133,6 +2134,7 @@ def deserialize(
             # we also need to bump unbacked sym[float,int] counters in the
             # shape env to accommodate unbacked symbols in the exported program
+            self.unbacked_symbols: set[sympy.Symbol] = set()
             count_unbacked_symfloat, count_unbacked_symint = -1, -1
```
I'll keep this as-is for now, symbol_name_to_range here contains the extra symbols. Gonna leave that as a separate cleanup job.

pianpwk committed:
[export] don't use unbacked_renamings in export (#147574)

Summary:
Plan: avoid the use of unbacked renamings, and introduce a pass run in `_produce_aten_artifact` that recomputes unbacked bindings. Decided to do this because in we don't serialize unbacked renamings (or any ShapeEnv state), so this used to compose poorly with de/serialization. This hopefully establishes the invariant that the unbacked binding keys are always in sync with the example values (i.e. same indices, and removed if the symbol is replaced / specialized).

Involved a refactor of compute_unbacked_bindings.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv


Reviewed By: avikchaudhuri

Differential Revision: D69961070

Pulled By: pianpwk
